### PR TITLE
Optimize 'aao' objc analysis ##anal

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

The aao command has been optimized a couple of times, based on the origin which was a Python script, the boost compared to the original script is more than 10 times faster. With this refactoring and cleanup I plan to make it even faster.

* [x] Reuse heap allocation between 
* [x] Use size_t instead of int for loops
* [x] Use bool ref to indicate success instead of UT64_MAX
* [x] Reduce unnecessary dereferences on non modified variables
* [x] Use more const and inline
* [x] Improve rowlog messages
* [x] Remove the use of sdb_fmt
* [x] Replace SDB+sprintf into plain HtUP+Vector
* [.] Use a single allocation instead of all the readQword calls (future optimization)

**Test plan**

There are objc tests for this already, but i may add another one and do proper benchmarking.

**Closing issues**

None

### CODE REVIEW

File: libr/core/anal_objc.c [hunk 1]
```diff
@@ -9,90 +9,137 @@
...
 	ut64 namePtr = readQword (objc, classMethodsVA, &readSuccess);
 	if (!readSuccess) {
 		return UT64_MAX;
 	}
 
-	int i, cnt = 0;
-	ut64 res_at = 0LL;
-	const char *k = addr_key (namePtr);
+	size_t i, cnt = 0;
+	ut64 ref = 0LL;
```
Comment (trufae) [inline libr/core/anal_objc.c:122]: better use UT64_MAX for the unset ref value

**Comment by trufae** [comment]
I've added more tests, more checks and the current status is like this (maybe it's wrong, so it needs more testing still):

before:

* aao find 355 xrefs in 7.5s 

now

* aao find 17095 in 6s

**Comment by trufae** [comment]
I think i will left the readQword optimization for later, the reads are done from objc_const and objc_data sections mainly

```
23  0x011b5670    0x1db0 0x1011b5670    0x1db0 -rw- 23.__DATA.__objc_const
24  0x011b7420     0x688 0x1011b7420     0x688 -rw- 24.__DATA.__objc_selrefs
25  0x011b7aa8      0x10 0x1011b7aa8      0x10 -rw- 25.__DATA.__objc_protorefs
26  0x011b7ab8     0x148 0x1011b7ab8     0x148 -rw- 26.__DATA.__objc_classrefs
27  0x011b7c00      0x28 0x1011b7c00      0x28 -rw- 27.__DATA.__objc_superrefs
28  0x011b7c28      0x50 0x1011b7c28      0x50 -rw- 28.__DATA.__objc_ivar
29  0x011b7c78     0x280 0x1011b7c78     0x280 -rw- 29.__DATA.__objc_data
```

So, i would appreciate some feedback and more testing before merging this.

File: libr/core/anal_objc.c [hunk 2]
```diff
@@ -11,142 +11,171 @@
...
-	const int objc2ClassSize = 0x28;
-	const int objc2ClassInfoOffs = 0x20;
-	const int objc2ClassMethSize = 0x18;
-	const int objc2ClassBaseMethsOffs = 0x20;
-	const int objc2ClassMethImpOffs = 0x10;
-
-	objc.core = core;
-	objc.word_size = (core->rasm->bits == 64)? 8: 4;
-
+static RCoreObjc *core_objc_new (RCore *core) {
```
Comment (ret2libc) [inline libr/core/anal_objc.c:172]: style

File: libr/core/anal_objc.c [hunk 3]
```diff
@@ -11,142 +11,171 @@
...
 		}
 	}
-	if (cnt > 1) {
-		return 0LL;
+	if (cnt > 1 || ref == 0 || ref == UT64_MAX) {
+		*rfound = false;
+		return UT64_MAX;
 	}
-	return res_at;
+	// (cnt > 1 && ref != 0);
```
Comment (ret2libc) [inline libr/core/anal_objc.c:126]: remove comment

**Comment by hexploitable** [comment]
### Test app 1

#### Old:
```
[0x100016740]> ?t aao
[x] Analyzing searching references to selref
[x] Parsing metadata in ObjC to find hidden xrefs
[x] A total of 74303 xref were found
[x] Set 2759 dwords at 0x10014a7b0
2.118805
```

#### New:
```
[0x100016740]> ?t aao
[x] Analyzing code to find selref references
[x] Parsing metadata in ObjC to find hidden xrefs
[x] Found 74303 objc xrefs in 2759 dwords.
1.800527
```

### Test app 2 (obfuscated)

#### Old:
```
[0x10005dd14]> ?t aao
[x] Analyzing searching references to selref
[x] Parsing metadata in ObjC to find hidden xrefs
[x] A total of 2016287 xref were found
[x] Set 16949 dwords at 0x1014e2bd8
49.050626
```

#### New:
```
[0x10005dd14]> ?t aao
[x] Analyzing code to find selref references
[x] Parsing metadata in ObjC to find hidden xrefs
[x] Found 2016359 objc xrefs in 16949 dwords.
38.828722
```

**Comment by trufae** [comment]
@ret2libc comments addressed, ready to go. there are still two more optimizations i plan to do in this code. But i don't want to make those improvements take that long to be merged and it's currently way better than it is in master

**Comment by XVilka** [comment]
@mrmacete could you please confirm? I think it is good to be merged before 4.5.0.

**Pull request merged**